Example of how to reject errors and display error messages in the error template

ianwdunlop opened the pull request:
Based on my experience with error handling in an ember app I think the guide needs to have an example of rejecting a promise and handling the error message.

trek commented:
Thanks for the submission. This doesn't match the code sample style used by the rest of the guides. Can you make them more like our other examples?

ianwdunlop commented:
I've changed the formatting in the code samples to make them look more like the current ones and added some path names to the files they may be in. If there is anything specific you are thinking of then just add a comment and I'll get it changed. Thanks.

trek commented on source/routing/loading-and-error-substates.md:
````diff
@@ -187,12 +187,73 @@ will look for:
+```
+
+```app/controllers/errorController.js
+App.ErrorController = Ember.ObjectController.extend({
````
No globals please.

ianwdunlop replied:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

Hello,

I'm guessing you want the Ember CLI type stuff like

export default Ember.Route.extend({

??

Brave new world, haven't even begun to get my head round it all yet....

Cheers,

Ian

On 30/03/15 14:59, Trek Glowacki wrote:
>> +This error "reason" will be passed to the error route available
>> to the app and +can then be set as the error controllers model
>> via the routes `setupController` method. You can then +display
>> specific error messages in the error template based on this error
>> "reason". You can +use the appropriate error controller eg.
>> `ErrorController` to display a message based on the +error
>> message properties. For example, + +```app/router.js 
>> +App.ErrorRoute = Ember.Route.extend({ +  setupController:
>> function(controller, model, params) { +
>> controller.set('model', model); +  } +}); +``` + 
>> +```app/controllers/errorController.js +App.ErrorController =
>> Ember.ObjectController.extend({
> 
> No globals please.
> 
> --- Reply to this email directly or view it on GitHub: 
> https://github.com/emberjs/guides/pull/141/files#r27391826
> 
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1

iQEcBAEBAgAGBQJVGViyAAoJEPK45GBX+Cy5zLQH/2BCZLe7HZ2FySbgNYaVQ1Q8
Ke7i5r4lhLKwgAtDtqYlmogI2smUx5ejExiFWAPeyILkH94rsoa49Hjm/RMeJuEo
/XeEEoJ50xwSCNq5oWtjG18S6DzZw0nlDSZYMZsItuQ647vCyFP5jfKwMEi8qAQu
TOrKhSBWgjHgjCJr88oeJMqkpImuoijmNofthcHf5bey0kJiamJwce+Hpqd5xbns
AUVoNjqtk7VkfRyX8AP2WXm+WVOdin47ewNfqPA+IKI4b+tlCfqF7vrRhGaUCera
CCe1eY3O4tw2lb6tk3XPZp9HMpXHCcouxTUqFhct+nhEmEFFbaMv30e8F6+BQJo=
=VpYA
-----END PGP SIGNATURE-----

trek replied:
Yes. Please use ES6 modules. The rest of the guides have already been converted.

ianwdunlop commented:
Removed globals for error controller and route and followed the ember cli conventions

locks commented on source/routing/loading-and-error-substates.md:
````diff
@@ -187,12 +187,73 @@ will look for:
+
+```app/routes/error.js
+export default Ember.Route.extend({
+  setupController: function(controller, model, params) {
````
This is the default behaviour, why are you explicitly declaring it?

ianwdunlop replied:
I can't recall exactly why but it may been an issue I was encountering with Ember at the time. The most important bit is the reject({"message": response.status}) part, without that no error message was bubbling up for me Happy for you to remove the controller setup part.

michaelrkn replied:
can you double-check and see whether this is necessary?

ianwdunlop replied:
Hello, just checked and you are correct that it is not needed.

locks commented on source/routing/loading-and-error-substates.md:
```diff
@@ -187,12 +187,73 @@ will look for:
+export default Ember.ObjectController.extend({
+  is404: function() {
+    return this.get('model.message') === 404;
+  }.property('model.message'),
```
Can you change the computed properties to this style please: 

```
is404: Ember.computed('model.message', function() {
  return this.get('model.message') === 404;
}),
```

ianwdunlop commented:
Changed my error example to use Ember.computed(..) rather than .property(...)

michaelrkn commented on source/routing/loading-and-error-substates.md:
````diff
@@ -187,12 +187,74 @@ will look for:
+promise with the HTTP response code as the error message:
+
+```js
+  reject({"message": response.status})
````
where is this code supposed to go?

michaelrkn commented:
Sorry for all the back and forth @ianwdunlop. I've been thinking this over for the past couple days. I feel like the code example in your PR is too specific - it's almost implying that it is the "right" way to handle errors - and complicated - the code is spread out through a few different examples and relatively long code blocks. I'd love to see this concept fleshed out, as you've started here, but I think there's a way to make it simpler and more general. If you're up to take another crack at it, by all means go for it! If not, I'll probably give it a shot in the next week or so. Also, maybe @locks can offer guidance and/or pick this up.

ianwdunlop commented:
I just based it on code I wrote to handle errors so it therefore meets my problem. For me it was about catching errors when fetching models. The initial problem for me was the vagueness in the guides which meant a lot of head scratching. The key bit was just figuring out that you have to send an object back with the reject in the promise. I guess you can make everything simpler but then it might end up back where I started!

michaelrkn commented:
I totally agree about the vagueness in the guides currently! There's actually an open issue now to improve it (#328). A few of us on the docs team had a chat about how to better get content input from the community generally, and we want to encourage people to write up a blog post on the topic before opening a PR to add big chunks of new content. That will get the content out there faster, and will help us maintain a consistent style throughout the guides. I think what you've written is a really great candidate for that process, and I'd love to see you turn it into a blog post that then could be considered for adaption into the guides. I'm going to close this PR for now, but could you comment with a link to the blog post on #328 so that we can add that to the general thinking about how to improve this? Thank you!
